assets/approval.py

```diff
@@ -0,0 +1,344 @@
+        on the dynamodb lock item.
+        :param target_dir: a temporary directory which will be exposed as an output
+        :param source: is the same value as passed to check
+        :param version: is the same type of value passed to check, and specifies the version to fetch.
+        :param params: is an arbitrary JSON object passed along verbatim from params on a get.
+        :return: a dict with the version fetched and the metadata of the lock
+        """
+        log.debug('source: %s', source)
+        log.debug('version: %s', version)
+
+        if not version:
+            version = {"timestamp": 0}
+
+        if params:
+            need_approval = params.get('need_approval', False)
+
+        # Does the get should wait for an approval or not ?
+        if 'lock_name' in params and 'need_approval' in params:
+            approval_lock = self.engine.query(Approval) \
+                .filter(
+                    pool=self.pool,
+                    lockname=params.get('lock_name')) \
+                .index('ts-index') \
+                .first(desc=True)
+
+            # We want to wait until the approve is done
+            if approval_lock.need_approval:
+                while approval_lock.approved is None:
+                    log.info("The lock %s is waiting for an approval" % params['lock_name'])
+                    # Query the lock item in the loop
+                    refresh_approval = self.query_lock(params['lock_name'])
+                    # Is it approved ?
+                    if refresh_approval.approved:
+                        approval_lock.approved = True
+                    # If not, we should fail the job and release the lock
+                    if refresh_approval.approved is False:
+                        log.info("The lock hasn't been approved, exiting")
+                        approval_lock.claimed = False
+                        approval_lock.approved = None
+                        self.engine.save(approval_lock, overwrite=True)
+                        exit(1)
+                    else:
+                        # Is hasn't been approved or rejected, waiting a bit more
+                        time.sleep(self.wait_lock)
+        else:
+            # There is no approval, we have just a normal lock. Let's fetch the lock
+            approval_lock = self.engine.query(Approval)\
+                .filter(
+                    Approval.timestamp >= datetime.fromtimestamp(Decimal(version.get('timestamp'))),
+                    pool=self.pool)\
+                .index('ts-index')\
+                .first(desc=True)
+
+        metadata = []
+        if approval_lock:
+            for key in approval_lock.keys_():
+                value = getattr(approval_lock, key)
+                if type(value) is datetime:
+                    value = str(Decimal(value.timestamp()))
+                if type(value) is bool:
```

Comment:
maybe `elif` to avoid to evaluate the statement again in case of the first one match